res/controllers/M-Audio-Xponent-Advanced-scripts.js

```diff
@@ -431,60 +473,57 @@ MaudioXponent.effectButton = function(channel, control, value, status, group) {
+            if (focusedEffect === button) {
+                // Toggle focused Effect
+                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + focusedEffect + "]";
+                var enabled = !engine.getValue(group, "enabled");
+                engine.setValue(group, "enabled", enabled);
+            } else {
+                // Focus effect
+                engine.setValue(group, "focused_effect", button);
+                engine.setValue(group, "show_focus", true);
+            }
+        }
     }
 };
 
-MaudioXponent.effectParameter = function(channel, control, value, status, group) {
-    // script.midiDebug(channel, control, value, status, group);
+MaudioXponent.effectKnob = function(channel, control, value) {
     var knob = control - 0x0B;
     var scaledValue = value / 0x7F;
 
-    if (MaudioXponent.state.focusedEffectUnit != 0) {
+    if (MaudioXponent.state.focusedEffectUnit === 0) {
         if (MaudioXponent.rightDeck.shift) {
-            // Manipulate focused EffectUnit parameters
-            if (knob == 4) {
-                // Wet / Dry
-                engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]", "mix", scaledValue);
-            } else {
-                // Other parameter
-                engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect1]", "parameter" + knob, scaledValue);
-            }
+            // Manipulate Wet/Dry mix
+            engine.setParameter("[EffectRack1_EffectUnit" + knob + "]", "mix", scaledValue);
         } else {
-            // Manipulate focused EffectUnit meta/super knobs
-            if (knob == 4) {
-                // Super-knob
-                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]";
-                engine.setParameter(group, "super1", scaledValue);
-            } else {
-                // Meta-knob
-                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + knob + "]";
-                engine.setParameter(group, "meta", scaledValue);
-            }
+            // Manipulate Unit Super-knobs
+            engine.setParameter("[EffectRack1_EffectUnit" + knob + "]", "super1", scaledValue);
         }
     } else {
-        if (MaudioXponent.rightDeck.shift) {
-            // Manipulate Wet/Dry mix
-            group = "[EffectRack1_EffectUnit" + knob + "]";
-            engine.setParameter(group, "mix", scaledValue);
+        // An effect has focus
+        var focusedEffect = engine.getValue("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]", "focused_effect");
+        
+        if (focusedEffect === 0) {
+            // Manipulate focused EffectUnit meta knobs
+            engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + knob + "]", "meta", scaledValue);
```

Comment:
That's actually what I had it doing before, but two different methods of getting to the same control seemed confusing. Can we say for sure that no skin will show more than three effects in a unit? Is three considered a hard limit? If so, then the fourth knob could definitely be used for the mix. I just don't want to find that in the final release there are skins with four effects per unit, and my mapping makes no sense when used with them.

I'd personally love it if Mixxx gets detachable windows in the future. I live in a multi-monitor world all day long, and I can totally see library and moving it to a second screen which would leave more room on the primary monitor for things like effects. With more screen, you could have more effects showing. I'm okay with saying that the mapping can only handle the first four because that's more than adequate. If you need more than that, then maybe it's time to start considering external hardware.